share/availability/light/availability.go

```diff
@@ -75,62 +77,80 @@ func (la *ShareAvailability) SharesAvailable(ctx context.Context, header *header
 	}
 	defer release()
 
-	// load snapshot of the last sampling errors from disk
 	key := datastoreKeyForRoot(dah)
+	samples := &SamplingResult{}
+
+	// Attempt to load previous sampling results
 	la.dsLk.RLock()
-	last, err := la.ds.Get(ctx, key)
+	data, err := la.ds.Get(ctx, key)
 	la.dsLk.RUnlock()
-
-	// Check for error cases
-	var samples []Sample
-	switch {
-	case err == nil && len(last) == 0:
-		// Availability has already been validated
-		return nil
-	case err != nil && !errors.Is(err, datastore.ErrNotFound):
-		// Other error occurred
-		return err
-	case errors.Is(err, datastore.ErrNotFound):
-		// No sampling result found, select new samples
-		samples = selectRandomSamples(len(dah.RowRoots), int(la.params.SampleAmount))
-	default:
-		// Sampling result found, unmarshal it
-		err = json.Unmarshal(last, &samples)
+	if err != nil {
+		if !errors.Is(err, datastore.ErrNotFound) {
+			return err
+		}
+		// No previous results; create new samples
+		samples = NewSamplingResult(len(dah.RowRoots), int(la.params.SampleAmount))
+	} else {
+		err = json.Unmarshal(data, samples)
 		if err != nil {
 			return err
 		}
+		// Verify total samples count.
+		totalSamples := len(samples.Remaining) + len(samples.Available)
+		if totalSamples != int(la.params.SampleAmount) {
```

Comment:
The fact that this param is configurable doesn't work with this error. We could reconsider the need for this param, but that's separate.